pkg/util/push/otel.go

```diff
@@ -61,14 +63,33 @@ func OTLPHandler(
 			}
 
 		default:
-			return nil, fmt.Errorf("unsupported content type: %s, supported: [%s, %s]", contentType, jsonContentType, pbContentType)
+			return nil, httpgrpc.Errorf(415, "unsupported content type: %s, supported: [%s, %s]", contentType, jsonContentType, pbContentType)
 		}
 
 		if r.ContentLength > int64(maxRecvMsgSize) {
 			return nil, fmt.Errorf(messageSizeLargerErrFmt, r.ContentLength, maxRecvMsgSize)
 		}
 
-		reader := http.MaxBytesReader(nil, r.Body, int64(maxRecvMsgSize))
+		reader := r.Body
+		// Handle compression.
+		switch r.Header.Get("Content-Encoding") {
+		case "gzip":
+			gr, err := gzip.NewReader(reader)
+			if err != nil {
+				return nil, err
+			}
+			reader = gr
+
+		case "":
+			// No compression.
+
+		default:
+			return nil, httpgrpc.Errorf(415, "unsupported compression: %s. Only \"gzip\" or no compression supported", r.Header.Get("Content-Encoding"))
+		}
+
+		// Protect against a large input.
+		reader = http.MaxBytesReader(nil, reader, int64(maxRecvMsgSize))
```

Comment:
Can we detect error returned `MaxBytesReader` (see `util.IsRequestBodyTooLarge(err)`) and convert it to `http.StatusRequestEntityTooLarge`? (Unit test covering this logic would be good too)

Reply:
Thanks for the review, and yes, using the `http.` variables make things much more readable.

I've added tests for everything only to find that we don't use `httpgrpc` to unwind errors, so I did that now too.